File: src/main/java/fiit/hipstery/publisher/bl/service/ArticleService.java
package fiit.hipstery.publisher.bl.service;

import fiit.hipstery.publisher.dto.*;

import java.io.IOException;
import java.util.UUID;

public interface ArticleService {

    /**
     * @param id is unique Id of article
     * @param currentUser is unique Id of currently logged in user
     * @return detailed version of article
     */
    ArticleDetailedDTO getArticleById(UUID id, UUID currentUser) throws IOException;

    /**
     * Filters articles according to filter.
     * @param filterCriteria Object that contains filters, also contains bounds of articles
     * @param currentUser is unique Id of currently logged in user
     * @return array of simple versions of filtered articles within preset bounds
     */
    ArticleSimpleListDTO getArticles(FilterCriteria filterCriteria, UUID currentUser);

    /**
     * Fetch all articles in range | it is faster then using getArticles with filter
     * @param lowerIndex lower bound of desired articles
     * @param upperIndex upper bound of desired articles
     * @param currentUser is unique Id of currently logged in user
     * @return array of simple versions of articles within preset bounds
     */
    ArticleSimpleListDTO getArticlesInRange(int lowerIndex, int upperIndex, UUID currentUser);

    /**
     * Fetch all articles from desired author | it is faster then using getArticles with filter
     * @param authorId is unique Id of desired author
     * @param currentUserId is unique Id of currently logged in user
     * @param lowerIndex lower bound of desired articles
     * @param upperIndex upper bound of desired articles
     * @return array of simple versions of articles within preset bounds from desired author
     */
    ArticleSimpleListDTO getArticlesByAuthor(UUID authorId, UUID currentUserId, int lowerIndex, int upperIndex);

    /**
     * Fetch all articles from desired publisher | it is faster then using getArticles with filter
     * @param publisherId is unique Id of desired publisher
     * @param currentUserId is unique Id of currently logged in user
     * @param page Index of desired page
     * @param pageSize number of articles in one page
     * @return array of simple versions of articles within preset bounds from desired publisher
     */
    ArticleSimpleListDTO getArticlesByPublisher(UUID publisherId, UUID currentUserId, int page, int pageSize);

    /**
     * Inserts article into table of articles
     * @param article Detailed version of article
     * @return UUID of new article
     */
    String insertArticle(ArticleInsertDTO article);

    /**
     * Soft deletes desired article
     * @param articleId is unique Id of desired article
     * @return true if softdelete was successful
     */
    boolean deleteArticle(UUID articleId);

    /**
     * Insert comment into table of comments, every object of comment includes id of article to which it belongs to.
     * @param comment new comment
     */
    void insertComment(CommentInsertDTO comment);

    /**
     * Like article
     * @param articleId is unique Id of liked article
     * @param userId is unique Id of user
     * @return number of likes on article
     */
    int likeArticle(UUID articleId, UUID userId);

    /**
     * Unlike article
     * @param articleId is unique Id of liked article
     * @param userId is unique Id of user
     * @return number of likes on article
     */
    int unlikeArticle(UUID articleId, UUID userId);

    /**
     * @return Number of articles
     */
    int getNumberOfArticles();
}

File: src/web/src/app/service/article.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpHeaders, HttpParams} from '@angular/common/http';
import {ArticleDetailedDTO, ArticleSimpleListDTO} from '../dto/dtos';
import {ArticleSimpleDTO} from '../dto/dtos';
import {Observable} from 'rxjs';
import {environment} from '../../environments/environment';
import {SessionService} from './session.service';

@Injectable({
  providedIn: 'root'
})
export class ArticleService {

  constructor(private httpClient: HttpClient, private sessionService: SessionService) {
  }

  public getArticleById(id: string): Observable<ArticleDetailedDTO> {
    const headers = new HttpHeaders({
      'Auth-Token': this.sessionService.getSession()
    });

    return this.httpClient.get<ArticleDetailedDTO>(environment.ROOT_URL + '/article/id/' + id, {headers});
  }

  public getFilteredArticles(
    title?: string,
    author?: string,
    category?: string,
    publisher?: string,
    lowerIndex?: number,
    upperIndex?: number,
  ): Observable<ArticleSimpleListDTO> {
    const headers = new HttpHeaders({
      'Auth-Token': this.sessionService.getSession()
    });

    let params = new HttpParams();
    if (title && title !== '') {
      params = params.append('title', title);
    }
    if (author && author !== '') {
      params = params.append('author', author);
    }
    if (category && category !== '') {
      params = params.append('category', category);
    }
    if (publisher && publisher !== '') {
      params = params.append('publisher', publisher);
    }
    if (lowerIndex) {
      params = params.append('lowerIndex', String(lowerIndex));
    }
    if (upperIndex) {
      params = params.append('upperIndex', String(upperIndex));
    }

    return this.httpClient.get<ArticleSimpleListDTO>(environment.ROOT_URL + '/article', {headers, params});
  }

  public getArticlesInRange(lower: number, upper: number): Observable<ArticleSimpleListDTO> {
    const headers = new HttpHeaders({
      'Auth-Token': this.sessionService.getSession()
    });

    return this.httpClient.get<ArticleSimpleListDTO>(
      environment.ROOT_URL + '/article/index/' + lower + '-' + upper, {headers});
  }

  public likeArticle(articleId: string): Observable<number> {
    const headers = new HttpHeaders({
      'Auth-Token': this.sessionService.getSession()
    });

    return this.httpClient.put<number>(environment.ROOT_URL + '/article/like/' + articleId, null, {headers});
  }

  public unlikeArticle(articleId: string): Observable<number> {
    const headers = new HttpHeaders({
      'Auth-Token': this.sessionService.getSession()
    });

    return this.httpClient.put<number>(environment.ROOT_URL + '/article/unlike/' + articleId, null, {headers});
  }
}
